### DISCUSSION TITLE
Windows printing alignment issue

### DISCUSSION THREAD

**Discussion by DanyallAslam** [microsoft/Dynamics365Commerce.Solutions#31] [Q&A] [answered]
X and Z report is cutting off, how do we fix it while using windows printer for printing.

![Uploading image.png…]()

**Comment by madyke** [comment]
Hi @DanyallAslam - it looks like the image you tried uploading has a broken link. Can you try uploading again so we can see what's getting cut off?

**Reply by DanyallAslam** [reply]
> Hi @DanyallAslam - it looks like the image you tried uploading has a broken link. Can you try uploading again so we can see what's getting cut off?

![image](https://user-images.githubusercontent.com/125641838/220526029-208e9230-f77b-4677-b127-3c78a296f4c2.png)

**Reply by madyke** [reply]
Hi @DanyallAslam - This GitHub forum is focused on development-related issues and questions, but from the photo provided, this looks like more of a configuration issue. Please raise a support ticket to get assistance on this problem.

**Reply by gilcardoso-msft** [reply]
Hi @madyke, the issue is that @DanyallAslam was using the HWS WindowsPrinterSample sample, from RetailSDK, in order to control the margins. However, now that they are switching to CommerceSDK, that option is not availabe, and they are struggling to upgrade/merge the sample to CommerceSDK as several references are not available from the Nugets. Any suggestions?

**Reply by madyke** [reply]
Thanks @gilcardoso-msft and @DanyallAslam for the extra details. We have a task in our backlog to migrate some of the key samples from RetailSDK to Commerce SDK, but we just haven't had the bandwidth to start that move yet.

For Hardware Station, the changes that are a part of Commerce SDK essentially boil down to removing a bunch of the contracts that were published in RetailSDK. For RetailSDK, HWS did not really offer true extensibility, and so our mitigation was to provide customers with more freedom (and sometimes even copies of our source code) so that they could replicate the core functionality in their extensions. Commerce SDK is working to provide true extensibility, just like in POS, but as you've discovered, that entails locking down the old contracts and instead providing specific extension points for the customer to use.

We're still working on the full extensibility story for HWS, but in the meantime, if you can provide a list of what references you are missing after the migration, I can pass that list along to our HWS extensibility expert, who can provide possible recommendations and workarounds.

**Reply by DanyallAslam** [reply]
Microsoft.Dynamics.Commerce.HardwareStation.Peripherals.DefaultLogo      

Microsoft.Dynamics.Commerce.HardwareStation.Peripherals.OposDeviceManager

Microsoft.Dynamics.Commerce.HardwareStation.Peripherals.OposHelper       

Microsoft.Dynamics.Commerce.HardwareStation.Peripherals.TextPart         

Microsoft.Dynamics.Commerce.HardwareStation.Peripherals.TextLogoParser   

Microsoft.Dynamics.Commerce.HardwareStation.Peripherals.TextType

**Reply by gilcardoso-msft** [reply]
Hi @madyke, any updates on this? Is any additional info necessary besides the list provided by @DanyallAslam in the previous message? Thanks

**Reply by madyke** [reply]
Hello @gilcardoso-msft and @DanyallAslam - apologies for the delay here. I have confirmed with our HWS extensibility expert that these missing references are due to the design changes I mentioned above, and that they will be replaced in Commerce SDK with new extension points rather than providing customers direct access to the packages themselves.

A new set of samples for HWS extensibility should be published soon that will demonstrate the new workflow pattern for how customers can accomplish these scenarios. Once they are live, I will let you know.

**Reply by hanhnguyen0288** [reply]
Hello @madyke , I am aslo interested in this topic. May I know if there is any updates on this. Do you have any estimated time for when new set of samples will be published. Thank you!

**Reply by madyke** [reply]
Hi @hanhnguyen0288 - I confirmed with our HWS extensibility expert that the samples are finished and we are working on backporting them to 10.0.33. They should hopefully be available within the next week.

**Reply by hanhnguyen0288** [reply]
> Hi @hanhnguyen0288 - I confirmed with our HWS extensibility expert that the samples are finished and we are working on backporting them to 10.0.33. They should hopefully be available within the next week.

That's great. Thank you @madyke.

**Reply by madyke** [reply]
Hi folks - sorry for the delay here, we ran into an issue with backporting the samples to 10.0.33 because some of them used a new API that was not available in that version. We are instead working on backporting them to 10.0.34, and hope to have them released in public preview next week.

**Reply by DanyallAslam** [reply]
Still having same issue 

![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/125641838/bdedda9d-8737-455e-b49d-19faca44dc25)

**Reply by DanyallAslam** [reply]
@madyke  Still waiting for this issue

**Reply by madyke** [reply]
@DanyallAslam & @hanhnguyen0288 - apologies for the slow response, Luke and I had both been pulled into some other escalations recently.

The updated HWS samples should have been published a few weeks ago as a part of the InStore samples:
[Dynamics365Commerce.InStore/src/HardwareStationSamples](https://github.com/microsoft/Dynamics365Commerce.InStore/tree/release/9.44/src/HardwareStationSamples)

There should be examples in those samples that provide details on how the new workflow pattern works in Commerce SDK. From the original question on this thread, the CustomOposDevicePrinter is probably a good starting point to take a look at:
[src/HardwareStationSamples/PeripheralCustomizations/Printer/CustomOposPrinterDevice.cs](https://github.com/microsoft/Dynamics365Commerce.InStore/blob/release/9.44/src/HardwareStationSamples/PeripheralCustomizations/Printer/CustomOposPrinterDevice.cs)

If you have any questions about the samples or still aren't sure what to do for your specific scenario, please let us know.

**Reply by DanyallAslam** [reply]
@madyke  we are using windows driver for printing, and the solution you suggest us is for OPOS driver Printer

**Reply by madyke** [reply]
Hi @DanyallAslam - We don't have a dedicated Windows printer sample, but the OPOS printer sample shows the general setup for how to create the handler and what kinds of methods and requests need to be implemented. Windows printers will be similar, but actually have some benefits because you can call the native APIs directly rather than having to call our internal OPOS APIs that OPOS printers need to use.

You should be able to use the OPOS sample as a starting point for the initial implementation, and can also reference the public .NET printer sample for developing your own Windows printer handler:
[PrintDocument Class Examples](https://learn.microsoft.com/en-us/dotnet/api/system.drawing.printing.printdocument?view=dotnet-plat-ext-7.0&viewFallbackFrom=netstandard-2.0)

**Reply by madyke** [reply]
Hi everyone, we have plans to migrate as many of the existing Retail SDK samples to the Commerce SDK as possible, but it will take time to get them all out. But since we've had more people chiming in on this thread and reaching out to CSS saying that they need this specific sample, we will be moving the Windows Printer sample up on our priority list. I'll be working with our HWS expert to get the sample updated and will update this post when we have something ready to share.

**Reply by spotityza** [reply]
Any update on the progress of this?

**Reply by madyke** [reply, marked as answer]
@spotityza @DanyallAslam @hanhnguyen0288 

Hi folks, it looks like the backport of the Windows printer sample was published today. You can find all of the [HWS samples here](https://github.com/microsoft/Dynamics365Commerce.InStore/tree/release/9.46/src/HardwareStationSamples), with instructions on how to set them up and test them. And the [Windows printer sample code is here](https://github.com/microsoft/Dynamics365Commerce.InStore/blob/release/9.46/src/HardwareStationSamples/PeripheralCustomizations/Printer/CustomWindowsPrinterDevice.cs). The sample demonstrates how you can modify the receipt content and formatting by creating a new request handler that will intercept the print request, giving you access to the string content of the receipt to be printed. This content can be modified and then sent via a new request to the next request handler, which contains all of the core code that is responsible for actually printing the receipt. If you need to reformat your receipt so that it will fit on different sized printers, to add or remove content from specific receipt types, or change the layout of the text in the receipt, then this sample should provide a basic implementation that you can build your specific requirements on top of.

The sample doesn't provide details on how to configure printer properties such as font size or margins due to design changes in what components of the printing model are extendible in Commerce SDK. After reviewing a lot of the business scenarios we've heard customers are trying to implement using Windows printer extensions, we've found that a lot of those scenarios are now supported through other extension capabilities, such as modifying receipt designs in HQ or modifying receipt text as the new sample demonstrates.

If you have a business scenario that you're trying to implement and you don't think that the newly published sample shows you what you need, please let us know and provide details on what the actual scenario is and why you want to implement it. We will do our best to help find whether there is another way that you can accomplish that without requiring you to write a bunch of custom Windows printing code. We are also starting to collect requirements for what customers would want if we provided printer configuration options such as the ability to control font size or margins, so if you have requirements for things like that this is a good chance to raise them so that they can be included in future design conversations.

We are also looking into providing a sample on how to create a receipt with a QR code in the future, so if you have requirements for that they should be addressed later down the road.

**Comment by DanyallAslam** [comment]


**Reply by madyke** [reply]
Hi @DanyallAslam - we have not migrated the dual display samples to Commerce SDK yet. They are in our backlog and we are currently hoping to get them released as a part of 10.0.36 (release dates are not set in stone and may slip if other priority work comes up).

**Comment by DanyallAslam** [comment]
Hi Matt,

We are looking for a WindowsPrinter device sample not an OPOSPrinter sample.

Thanks,
D

**Comment by spotityza** [comment]
Any update on this @madyke

**Comment by MuhammadShahrukhMasood** [comment]
Hi Support,

We need cash drawer sample in commercesdk on urgent basis, please entertain ASAP.

**Reply by spotityza** [reply]
Log it as a new discussion, it won't get visibility here
